File: mshc/catalog.pl
#!/usr/bin/perl
#
# Read all html files in the current directory, read the meta Microsoft.Help.Id, and build a catalog from it

use utf8;
use encoding 'UTF-8';
use Data::Dumper;
use XML::Simple;

binmode STDOUT, ':utf8';

sub shellesc($) {
    my $x = shift @_;

    $x =~ s/\'/\\\'/;
    $x =~ s/\"/\\\"/;
    $x =~ s/\$/\\\$/;

    return $x;
}

my @catalog;

open(CAT,">catalog.txt") || die;
binmode(CAT,':utf8');

open(LS,"find |") || die;
while (my $path = <LS>) {
    chomp $path;
    $path =~ s/^\.\///;
    next unless -f $path;
    next unless $path =~ m/\.(html|htm)$/i;

    print "Parsing $path...\n";
    $parser = new XML::Simple;

    my $html;
    open(X,"<$path") || die;
    binmode(X,':utf8');
    read(X,$html,1024*1024*128);
    close(X);

    utf8::encode($html);
    $data = XMLin($html);

    # we're looking for root -> head -> <meta name="Microsoft.Help.Id" content="...">
    $x = $data->{"head"};
    next unless defined($x);
    $x = $x->{"meta"};
    next unless defined($x);
    next unless ref($x) eq 'ARRAY';

    my $id = undef;
    my @a = @{$x};

    for ($i=0;$i < @a;$i++) {
        my $yy = $a[$i];
        my %x = %{$yy};

        my $name = $x{"name"};
        my $content = $x{'content'};

        next unless defined($name);
        next unless defined($content);

        if (lc($name) eq lc("Microsoft.Help.Id")) {
            print CAT "$path\n";
            print CAT "-id:$content\n";
            print CAT "\n";
        }
    }
}
close(LS);
close(CAT);

File: mshc/catalog.cpp

#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>

#include <libxml/parser.h>
#include <libxml/HTMLparser.h>

#include <string>

using namespace std;

static void chomp(char *s) {
    char *e = s + strlen(s) - 1;

    while (e > s && (*e == '\r' || *e == '\n')) *e-- = 0;
}

FILE *cat_fp = NULL;
FILE *ls_fp = NULL;

void GetHelpId(const char *path) {
    xmlNodePtr html=NULL,head=NULL,meta=NULL;
    xmlDocPtr doc;

    /* NTS: MSHC are generally XHTML */

    doc = xmlReadFile(path,NULL,XML_PARSE_RECOVER|XML_PARSE_NONET);
    if (doc == NULL) return;

    /* look for the HTML -> HEAD -> META entries */
    html = xmlDocGetRootElement(doc);
    while (html != NULL) {
        if (!xmlStrcmp(html->name,(const xmlChar*)"html"))
            break;

        html = html->next;
    }

    if (html != NULL) {
        head = html->children;
        while (head != NULL) {
            if (!xmlStrcmp(head->name,(const xmlChar*)"head"))
                break;

            head = head->next;
        }
    }

    if (head != NULL) {
        meta = head->children;
        while (meta != NULL) {
            if (!xmlStrcmp(meta->name,(const xmlChar*)"meta")) {
                string name,content;
                xmlChar *xp;

                xp = xmlGetNoNsProp(meta,(const xmlChar*)"name");
                if (xp != NULL) {
                    name = (char*)xp;
                    xmlFree(xp);
                }

                xp = xmlGetNoNsProp(meta,(const xmlChar*)"content");
                if (xp != NULL) {
                    content = (char*)xp;
                    xmlFree(xp);
                }

                if (name == "Microsoft.Help.Id") {
                    fprintf(cat_fp,"%s\n",path);
                    fprintf(cat_fp,"-id:%s\n",content.c_str());
                    fprintf(cat_fp,"\n");
                }
            }

            meta = meta->next;
        }
    }

    /* done */
    xmlFreeDoc(doc);
}

int main() {
    char line[1024];

    ls_fp = popen("find","r");
    if (ls_fp == NULL) return 1;

    cat_fp = fopen("catalog.txt","w");
    if (cat_fp == NULL) return 1;

    while (!feof(ls_fp) && !ferror(ls_fp)) {
        if (fgets(line,sizeof(line)-1,ls_fp) == NULL) break;
        chomp(line);

        /* must be .htm or .html */
        char *ext = strrchr(line,'.');
        if (ext == NULL) continue;

        if (!strcmp(ext,".htm") || !strcmp(ext,".html")) {
            printf("Reading %s\n",line);
            GetHelpId(line);
        }
    }

    fclose(cat_fp);
    fclose(ls_fp);
    return 0;
}

